flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolImpl.java

```diff
@@ -715,6 +717,19 @@ public boolean releaseTaskManager(final ResourceID resourceId, final Exception c
 		}
 	}
 
+	@Override
+	public SlotSnapshot createSlotSnapshot(ResourceID taskManagerId) {
+		Set<AllocatedSlot> totalSlots = new HashSet<>();
+		totalSlots.addAll(availableSlots.getSlotsForTaskManager(taskManagerId));
+		totalSlots.addAll(allocatedSlots.getSlotsForTaskManager(taskManagerId));
```

Comment:
No need to create the extra collection here:
```
@Override
	public SlotSnapshot createSlotSnapshot(ResourceID taskManagerId) {
		final Set<AllocatedSlot> availabeSlotsForTaskManager = availableSlots.getSlotsForTaskManager(taskManagerId);
		final Set<AllocatedSlot> allocatedSlotsForTaskManager = allocatedSlots.getSlotsForTaskManager(taskManagerId);

		List<OfferedSlot> offeredSlots = new ArrayList<>(availabeSlotsForTaskManager.size() + allocatedSlotsForTaskManager.size());
		for (AllocatedSlot allocatedSlot : Iterables.concat(availabeSlotsForTaskManager, allocatedSlotsForTaskManager)) {
			offeredSlots.add(new OfferedSlot(allocatedSlot.getPhysicalSlotNumber(), allocatedSlot.getAllocationId()));
		}
		return new SlotSnapshot(jobId, offeredSlots);
	}
```

Reply:
Good idea